### PR TITLE
Added Support preview Team Sync API for GitHub Enterprise Cloud users

### LINKED ISSUE #1200: Support preview Team Sync API for GitHub Enterprise Cloud users
GitHub Developer API announcement:
https://developer.github.com/changes/2019-06-12-team-sync/

This would be a great PR for any new contributor to this repo or a new Go developer.
All contributions are greatly appreciated!

Feel free to volunteer for any issue and the issue can be assigned to you so that others don't attempt to duplicate the work.

Please check out our [CONTRIBUTING.md](https://github.com/google/go-github/blob/master/CONTRIBUTING.md) guide to get started.

Thank you!

### PR DESCRIPTION
**Pull request by vaibhavsingh97**
Fixes #1200

### CODE REVIEW

File: github/github.go [hunk 1]
```diff
@@ -147,6 +147,9 @@ const (
 
 	// https://developer.github.com/changes/2019-04-11-pulls-branches-for-commit/
 	mediaTypeListPullsOrBranchesForCommitPreview = "application/vnd.github.groot-preview+json"
+
+	// https://developer.github.com/changes/2019-06-12-team-sync/
+	mediaTypeTeamSync = "application/vnd.github.team-sync-preview+json"
```
Comment (gmlewis) [inline github/github.go]: Can you please add `Preview` to the name to make it `mediaTypeTeamSyncPreview` ?

File: github/teams.go [hunk 2]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
 
 	return s.client.Do(ctx, req, nil)
 }
+
+// GroupObject reprents list of Idp Group.
```
Comment (gmlewis) [inline github/teams.go]: Please change this to: `// IDPGroupList represents a list of external identity provider (IDP) groups.`
I know GitHub spells it `IdP`, but Go always uses `ID` for "identification", so let's stick to the Go standard here too.
:smile:

File: github/teams.go [hunk 3]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+type GroupObject struct {
+	Groups []*Group `json:"groups,omitempty"`
+}
+
+// Group represents an Idp group status.
+type Group struct {
```
Comment (gmlewis) [inline github/teams.go]: Please change this to: `// IDPGroup represents an external identity provider (IDP) group.`

File: github/teams.go [hunk 4]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+	GroupID          *string `json:"group_id,omitempty"`
+	GroupName        *string `json:"group_name,omitempty"`
+	GroupDescription *string `json:"group_description,omitempty"`
+}
+
+// ListIDPGroupsInOrganization list IdP groups available in an organization.
```
Comment (gmlewis) [inline github/teams.go]: Please change `list IdP` to `lists IDP`.

File: github/teams.go [hunk 5]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+//
+// Github API docs: https://developer.github.com/v3/teams/team_sync/
```
Comment (gmlewis) [inline github/teams.go]: Please use the official capitalization for `GitHub`.

URL should be: https://developer.github.com/v3/teams/team_sync/#list-idp-groups-in-an-organization

File: github/teams.go [hunk 6]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// TODO: remove custom Accept header when this API fully launches.
+	req.Header.Set("Accept", mediaTypeTeamSync)
+
+	groups := new(GroupObject)
+	resp, err := s.client.Do(ctx, req, groups)
+
```
Comment (gmlewis) [inline github/teams.go]: Please delete this blank line between the `err` and its test.

File: github/teams.go [hunk 7]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+	if err != nil {
+		return nil, resp, err
+	}
+	return groups, resp, nil
+}
+
+// ListIDPGroupsForTeam list IdP groups connected to a team on GitHub.
+//
+// Github API docs: https://developer.github.com/v3/teams/team_sync/
```
Comment (gmlewis) [inline github/teams.go]: s/Github/GitHub/

URL should be: https://developer.github.com/v3/teams/team_sync/#list-idp-groups-for-a-team

File: github/teams.go [hunk 8]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+	if err != nil {
+		return nil, resp, err
+	}
+	return groups, resp, err
+}
+
+// CreateOrUpdateIDPGroupConnections creates, updates, or removes a connection between a team
+// and an IdP group.
+//
+// Github API docs: https://developer.github.com/v3/teams/team_sync/
```
Comment (gmlewis) [inline github/teams.go]: s/Github/GitHub/

URL should be: https://developer.github.com/v3/teams/team_sync/#create-or-update-idp-group-connections

File: github/teams.go [hunk 9]
```diff
@@ -476,3 +476,88 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
+	groups := new(GroupObject)
+	resp, err := s.client.Do(ctx, req, groups)
+	if err != nil {
+		return nil, resp, err
+	}
+	return groups, resp, err
+}
+
+// CreateOrUpdateIDPGroupConnections creates, updates, or removes a connection between a team
+// and an IdP group.
```
Comment (gmlewis) [inline github/teams.go]: s/IdP/IDP/

File: github/teams.go (hunk 2 above)
Comment (vaibhavsingh97) [reply github/teams.go]: should I also change the name of struct to `IDPGroupList` ?

Comment (gmlewis) [reply github/teams.go]: Yes, please.

**Comment by vaibhavsingh97** [comment]
@gmlewis Done with changes, please review 😄

**Comment by gmlewis** [comment]
Thank you, @gauntface!
Merging.

**Pull request merged**